File: Python/Easy/3120_count_the_number_of_special_characters_i.py
"""
3120. Count the Number of Special Characters I

You are given a string word. A letter is called special if it appears both in
lowercase and uppercase in word. Return the number of special letters in word.

Problem Link: https://leetcode.com/problems/count-the-number-of-special-characters-i/

Examples:
    Example 1:
        Input: word = "aaAbcBC"
        Output: 3
        Explanation: The special characters are 'a', 'b', and 'c'.

    Example 2:
        Input: word = "abc"
        Output: 0
        Explanation: No character appears in uppercase.

    Example 3:
        Input: word = "abBCab"
        Output: 1
        Explanation: The only special character is 'b'.

Constraints:
    1 <= word.length <= 50
    word consists of only lowercase and uppercase English letters.
"""

import string


class Solution:
    def numberOfSpecialChars(self, word: str) -> int:
        """
        Collapse word to a set of distinct characters, then count lowercase
        letters whose uppercase form is also present.

        Time complexity:  O(n)
        Space complexity: O(1) — the set holds at most 52 letters.
        """
        char_set: set[str] = set(word)
        ans: int = 0
        for c in char_set:
            if c.islower() and c.upper() in char_set:
                ans += 1
        return ans

    def numberOfSpecialCharsAscii(self, word: str) -> int:
        """
        Same set membership idea, but iterate the 26 lowercase letters
        directly so the result is independent of set iteration order.

        Time complexity:  O(n)
        Space complexity: O(1)
        """
        s: set[str] = set(word)
        return sum(c in s and c.upper() in s for c in string.ascii_lowercase)


if __name__ == "__main__":
    solution = Solution()

    test_cases = [
        ("aaAbcBC", 3),
        ("abc", 0),
        ("abBCab", 1),
        ("a", 0),
        ("aA", 1),
        ("AaBbCc", 3),
    ]

    for i, (word, expected) in enumerate(test_cases, 1):
        assert solution.numberOfSpecialChars(word) == expected, f"set test {i} failed: {word!r}"
        assert solution.numberOfSpecialCharsAscii(word) == expected, f"ascii test {i} failed: {word!r}"

    print("All test cases passed!")

File: Golang/Easy/3120_count_the_number_of_special_characters_i.go
package main

import (
	"fmt"
	"math/bits"
)

// numberOfSpecialChars counts letters that appear both lower- and upper-cased.
//
// Map approach: record every rune seen, then for each lowercase letter check
// whether its uppercase counterpart (c-'a'+'A') is also present.
//
// Time complexity:  O(n)
// Space complexity: O(1) — at most 52 keys.
func numberOfSpecialChars(word string) int {
	s := make(map[rune]bool)
	for _, c := range word {
		s[c] = true
	}

	ans := 0
	for c := 'a'; c <= 'z'; c++ {
		if s[c] && s[c-'a'+'A'] {
			ans++
		}
	}
	return ans
}

// numberOfSpecialCharsBitmask packs presence into two 26-bit masks, one for
// lowercase and one for uppercase. A letter is special iff its bit is set in
// both, so the answer is the popcount of their AND.
//
// Time complexity:  O(n)
// Space complexity: O(1)
func numberOfSpecialCharsBitmask(word string) int {
	var lowerMask, upperMask int32

	for i := 0; i < len(word); i++ {
		ch := word[i]
		if ch >= 'a' && ch <= 'z' {
			lowerMask |= 1 << (ch - 'a')
		} else if ch >= 'A' && ch <= 'Z' {
			upperMask |= 1 << (ch - 'A')
		}
	}

	bothPresent := lowerMask & upperMask
	return bits.OnesCount32(uint32(bothPresent))
}

// numberOfSpecialCharsArrays mirrors the bitmask version with two boolean
// presence arrays, then counts indices flagged in both.
//
// Time complexity:  O(n)
// Space complexity: O(1)
func numberOfSpecialCharsArrays(word string) int {
	const numOfLetters = 26
	var hasLower, hasUpper [numOfLetters]bool
	for i := 0; i < len(word); i++ {
		ch := word[i]
		if ch >= 'a' && ch <= 'z' {
			hasLower[ch-'a'] = true
		} else if ch >= 'A' && ch <= 'Z' {
			hasUpper[ch-'A'] = true
		}
	}

	count := 0
	for i := 0; i < numOfLetters; i++ {
		if hasLower[i] && hasUpper[i] {
			count++
		}
	}
	return count
}

func main() {
	testCases := []struct {
		word     string
		expected int
	}{
		{"aaAbcBC", 3},
		{"abc", 0},
		{"abBCab", 1},
		{"a", 0},
		{"aA", 1},
		{"AaBbCc", 3},
	}

	for _, approach := range []struct {
		label string
		fn    func(string) int
	}{
		{"numberOfSpecialChars       ", numberOfSpecialChars},
		{"numberOfSpecialCharsBitmask", numberOfSpecialCharsBitmask},
		{"numberOfSpecialCharsArrays ", numberOfSpecialCharsArrays},
	} {
		fmt.Printf("--- %s ---\n", approach.label)
		for i, tc := range testCases {
			result := approach.fn(tc.word)
			status := "PASS"
			if result != tc.expected {
				status = "FAIL"
			}
			fmt.Printf("  Example %d | Input: %q | Expected: %d | Result: %d | [%s]\n",
				i+1, tc.word, tc.expected, result, status)
		}
	}
}
